segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/TableCompactor.java

```diff
@@ -0,0 +1,448 @@
+ * the entry will now live at a higher offset. Some of the Table Segment's attributes are also atomically updated as part
+ * of this operation (namely {@link TableAttributes#COMPACTION_OFFSET} and {@link TableAttributes#TOTAL_ENTRY_COUNT}).
+ * - These copied entries are not indexed as part of compaction. Similarly to normal updates, the {@link WriterTableProcessor}
+ * will pick them up and index them.
+ */
+@Slf4j
+@RequiredArgsConstructor
+class TableCompactor {
+    //region Members
+
+    @NonNull
+    private final TableWriterConnector connector;
+    @NonNull
+    private final IndexReader indexReader;
+    @NonNull
+    private final Executor executor;
+
+    //endregion
+
+    /**
+     * Determines if Table Compaction is required on a Table Segment.
+     *
+     * @param info The {@link SegmentProperties} associated with the Table Segment to inquire about.
+     * @return True if compaction is required, false otherwise.
+     */
+    boolean isCompactionRequired(SegmentProperties info) {
+        long startOffset = getCompactionStartOffset(info);
+        long lastIndexOffset = this.indexReader.getLastIndexedOffset(info);
+        if (startOffset + this.connector.getMaxCompactionSize() >= lastIndexOffset) {
+            // Either:
+            // 1. Nothing was indexed
+            // 2. Compaction has already reached the indexed limit.
+            // 3. Not enough "uncompacted" data - at least this.maxCompactLength must be accumulated to trigger a compaction.
+            return false;
+        }
+
+        long totalEntryCount = this.indexReader.getTotalEntryCount(info);
+        long entryCount = this.indexReader.getEntryCount(info);
+        long utilization = totalEntryCount == 0 ? 100 : MathHelpers.minMax(Math.round(100.0 * entryCount / totalEntryCount), 0, 100);
+        long utilizationThreshold = (int) MathHelpers.minMax(this.indexReader.getCompactionUtilizationThreshold(info), 0, 100);
+        return utilization < utilizationThreshold;
+    }
+
+    /**
+     * Calculates the offset in the Segment where it is safe to truncate based on the current state of the Segment and
+     * the highest copied offset encountered during an index update.
+     *
+     * This method is invoked from the {@link WriterTableProcessor} after indexing. Since compaction is loosely coupled
+     * with indexing, the {@link WriterTableProcessor} does not have too many insights into what has been compacted or not;
+     * it can only decide based on the current state of the Table Segment and what it has just indexed. As such:
+     * - If recently indexed Table Entries indicate they were copied as part of a compaction, then it is safe to truncate
+     * at the highest copied offset encountered (since they are copied in order, by offset). Everything prior to this
+     * offset is guaranteed not to exist in the index anymore.
+     * - If no recently indexed Table Entry indicates it was copied as a result of compaction, then it may not be safe to
```

Comment:
Correct. The truncation only is invoked if no other exceptions happen during the indexing iteration, so if we encounter such copied offsets, they must have been indexed already and the index no longer points to the original offsets. Since all these entries are copied in order of offsets, it is safe to truncate at the highest one we find).